### ISSUE TITLE
double conversions

### ISSUE THREAD

**Issue by lshems** [EdgeTX/edgetx#276]
When opening SIMULATOR for T16 with FILE option, and choosing OpenTX2.3.12 file, it gets converted correctly, including model names and input line names.

When closing the sim, and restarting, it will do the conversion for a second time, messing all up.

**Comment by raphaelcoeffic** [comment]
Now this is funny! Thx!

**Comment by raphaelcoeffic** [comment]
@lshems which simulator did you mean? I just tried with `simu`, and it works, as far as I can see. No second version. Is this with the simulator using Qt?

**Comment by lshems** [comment]
It's is with the simulator that is started with ./simulator24

**Comment by lshems** [comment]
Also, the message is a little bit vague before conversion, and you can't prevent it from starting. It says something like 'sdcard needs conversion' which is not what is going to happen. The models are converted Inthe otx file.

**Comment by raphaelcoeffic** [comment]
> Also, the message is a little bit vague before conversion, and you can't prevent it from starting. It says something like 'sdcard needs conversion' which is not what is going to happen. The models are converted Inthe otx file.

Ok, then this is a companion error, it is probably still using wrong directory names, or something like that. @rotorman could you please have a look at it?

**Comment by lshems** [comment]
I will post a screenshot later. Still having some coffee first.

To reproduce, take X9D+.otx from the closed issue on 'some Lua model functions...' put it as file to load on the simulator profile, and start the simulator. It will convert all correct. Close the simulator and repeat

**Comment by rotorman** [comment]
@lshems From the title, you are likely talking about #227. Cannot find any *.otx file there to test with?
Otherwise I have simulator24 running in Qt Creator under Ubuntu in Debug mode, so could try to figure it out.

**Comment by rotorman** [comment]
I guess you mean this dialog here (tested with X9D+ simulator24):
![DataLoadError](https://user-images.githubusercontent.com/21011587/122673440-c2690f80-d1d0-11eb-94c5-f025eedde3b3.png)
I see this stems from
https://github.com/EdgeTX/edgetx/blob/3472bbdc781a93ef87fc6273b255889da921860e/companion/src/simulation/simulatorwidget.cpp#L269-L274
need to figure out why ret is 0, it does not hit the previous else case though:
https://github.com/EdgeTX/edgetx/blob/3472bbdc781a93ef87fc6273b255889da921860e/companion/src/simulation/simulatorwidget.cpp#L264-L267

**Comment by lshems** [comment]
https://github.com/EdgeTX/edgetx/files/6680426/t16_23.zip

Load the Ubuntu simulator with the profile dialog, choose this file as file to load.

Start, conversion begins after an SD card message, close, and load again in the same way

**Comment by lshems** [comment]
Change .zip to .otx of course

**Comment by rotorman** [comment]
Have Jumper T16 now running in debug session in Qt Creator. Started with:
![Startup](https://user-images.githubusercontent.com/21011587/122675385-69ea4000-d1d9-11eb-881e-abb0dc60047a.png)

The following comes out of debug output:
[DebugOutput.txt](https://github.com/EdgeTX/edgetx/files/6682418/DebugOutput.txt)

**Comment by lshems** [comment]
Ok,but it starts the conversion correct no?. Just SI and SJ warnings, but the conversions seem ok.

**Comment by lshems** [comment]
And then, after closing, it tries to convert the (now already converted) file again

**Comment by rotorman** [comment]
Yes, I see that it tries to convert every time one launches the simulator.
![StorageWarning](https://user-images.githubusercontent.com/21011587/122676747-a28d1800-d1df-11eb-839b-c55178188eb8.png)

![Converting](https://user-images.githubusercontent.com/21011587/122676828-fac41a00-d1df-11eb-9448-c804cff3efb6.png)

I see also that before this happens, the execution in the Qt simu app is passed to sim library at:
https://github.com/EdgeTX/edgetx/blob/3472bbdc781a93ef87fc6273b255889da921860e/companion/src/simulator.cpp#L352
I wonder if it happens, as there is no way for the radio simulator library (not the simulator binary) to write back the converted otx file into the host filesystem.
Did this work with OTX 2.3?

**Comment by rotorman** [comment]
Interestingly the otx file that got converted by the simulator differs in the content, except the models.txt. All binary files have been touched. You can compare here with the original file from above:
[t16_23_converted.zip](https://github.com/EdgeTX/edgetx/files/6682548/t16_23_converted.zip)
Thus, the question is, why is a new conversion triggered. This should be irrespective of the simulator, and run similarly on physical hw/firmware.

**Comment by rotorman** [comment]
By looking at the converted radio.bin, one sees that the tag is still old (0xDB = 219):
![radio](https://user-images.githubusercontent.com/21011587/122680473-a6289b00-d1ef-11eb-8438-cb5dc1b1d5b5.png)

Also the models have the old tag, here exemplary model1.zip:
![model1](https://user-images.githubusercontent.com/21011587/122680502-bfc9e280-d1ef-11eb-9f4b-01ad3f0efe21.png)

Seems that Companion24/Simulator24 do not cleanly save the new format.

@elecpower would you have tips, or be able to help out? Thanks!

**Comment by elecpower** [comment]
From memory the Companion Simulator makes a copy of the model and settings to be simulated and any changes are discarded when the simulator is closed. The simu saves it changes.

**Comment by rotorman** [comment]
@lshems please test PR #286 - for me this solves the issue.

**Comment by lshems** [comment]
Ok, so help me a little, just getting around here. 

I have created an ubuntu according to the instructions. All fine. I can pull the head branch, also fine.
But how do I pull your PR?

**Comment by rotorman** [comment]
@lshems `clone --recursive -b SimuCompFolders https://github.com/rotorman/edgetx.git pr286test` will create a local folder pr286test for you.

Please keep in mind, as @raphaelcoeffic hinted at #286, although this fixes the issue with repeated conversions, it might go amok with Companion, as the new data structure of version 220 needs to be additionally added there. Currently Companion is aware of data structures up to 219 only. This means, to get also Companion working, more work is required.

**Comment by lshems** [comment]
I don't need companion,(yet).

A working simulator starting on an otx file would do for the moment. So without the reconversions.

Still a little puzzled on how that should work on a real radio, when doing the conversion.

**Comment by raphaelcoeffic** [comment]
> I don't need companion,(yet).
> 
> A working simulator starting on an otx file would do for the moment. So without the reconversions.
> 
> Still a little puzzled on how that should work on a real radio, when doing the conversion.

I see, but the simulator you're using is heavily based on companion, which is the issue at hand right now.

**Comment by lshems** [comment]
Just for my understanding, the simulator is depending on companion for defining the 'load settings' for the profile, or for much more?

Because I would reckon the simulator is more sort of a hardware emulator. Otherwise you would have to update all things on the firmware as well as for the simulator??

Just trying to graps the architecture to be able to understand dependancies.

**Comment by raphaelcoeffic** [comment]
> Just for my understanding, the simulator is depending on companion for defining the 'load settings' for the profile, or for much more?
> 
> Because I would reckon the simulator is more sort of a hardware emulator. Otherwise you would have to update all things on the firmware as well as for the simulator??
> 
> Just trying to graps the architecture to be able to understand dependancies.

It does some reading of the model/radio file to be able to setup the controls. Then it seems it tweaks  the files and repacks them (did not know).

**Comment by lshems** [comment]
I always had the feeling that opentx was spaghetti code, based on the magic numbers mayhem in Lua API.

This is another indication. I'll better keep far from it....:)

**Comment by raphaelcoeffic** [comment]
> I always had the feeling that opentx was spaghetti code, based on the magic numbers mayhem in Lua API.
> 
> This is another indication. I'll better keep far from it....:)

Let's say: it has a long history.

**Comment by pfeerick** [comment]
> I always had the feeling that opentx was spaghetti code, based on the magic numbers mayhem in Lua API.

Magic numbers? Are you implying there is magic involved in OTX? :-P

**Comment by lshems** [comment]
Well, it runs by chance, so ...

**Comment by raphaelcoeffic** [comment]
@lshems I believe this issue is fixed, right? Can you please confirm?

**Comment by lshems** [comment]
I can correctly read a otx file and it gets converted in the simulator.

**Comment by rotorman** [comment]
@lshems meaning, we can close the issue, correct?

**Comment by lshems** [comment]
Yes :)
